fortnite_api/material.py

```diff
@@ -0,0 +1,140 @@
+"""
+MIT License
+
+Copyright (c) 2019-present Luc1412
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+from __future__ import annotations
+
+from typing import Any, Dict, Generic, Optional, Tuple
+
+from .abc import Hashable
+from .asset import Asset
+from .enums import CosmeticCompatibleMode
+from .http import HTTPClientT
+from .utils import get_with_fallback
+
+__all__: Tuple[str, ...] = ('MaterialInstanceImages', 'MaterialInstanceColors', 'MaterialInstance')
+
+
+class MaterialInstanceImages(Generic[HTTPClientT]):
+    """
+    .. attributetable:: fortnite_api.MaterialInstanceImages
+
+    Represents some images of a Material instance, as they are rendered in game.
+
+    Attributes
+    ----------
+    offer_image: :class:`fortnite_api.Asset`
+        The offer image of the Material instance. This is the image that is shown in the item shop.
+    texture: Optional[:class:`fortnite_api.Asset`]
+        The texture of the Material instance. This is the texture of the material instance.
+    background: :class:`fortnite_api.Asset`
+        The background of the Material instance. This is the background gradient of the material instance.
+    """
+
+    __slots__: Tuple[str, ...] = ('offer_image', 'background', 'texture')
+
+    def __init__(self, *, data: Dict[str, Any], http: HTTPClientT) -> None:
+        self.offer_image: Asset[HTTPClientT] = Asset(url=data['OfferImage'], http=http)
+
+        _texture = data.get('Texture')
+        self.texture: Optional[Asset[HTTPClientT]] = _texture and Asset(url=_texture, http=http)
+
+        _background = data.get('Background')
+        self.background: Optional[Asset[HTTPClientT]] = _background and Asset(url=_background, http=http)
```

Comment:
All keys are going to resolve to `Assets`, so it may be best to just transform all keys into `Mapping[str, Asset]` then (except `offer_image`)? I believe that trying to keep the dynamic keys as attributes may lead to issues down the road.

Reply:
Maybe we can type soft-type the dict. At least `OfferImage` and the manually supplied  Background could be more easily accessible, since they as used the most.

Reply:
Could you elaborate on what you mean by "type soft-type the dict"? I'm thinking that we basically transform the class to something like this:
```python
class MaterialInstanceImages(...):
    def __init__(self, *, data, http):
        self.offer_image: ... = ...
        self.raw: Dict[str, Asset[...]] = ...

    @property 
    def background(self) -> Optional[Asset[...]]: ...
 ```
 `offer_image` would be the only concrete attribute because it's manually supplied and the other keys that must be easily accessible will be properties. Additionally, `raw` will contain the parsed dynamic keys
 
 What do you think about this approach?

Reply:
`OfferImage` isn't manually supplied, but it's the most useful out of all. `Background` is created and passed by Fortnite-API.com and contains the OfferImage with the generated background. Don't understand, that `background` is a prop, but `offer_image` a attribute.

Maybe we implement `__getitem__` for accessing other keys and providing the most common ones as an attribute or prop.